File: vm/tests/windows_services.cpp
#include "../master.hpp"

static void check(bool condition, const char* message) {
  if (!condition) {
    std::cerr << message << std::endl;
    ::exit(1);
  }
}

int wmain(int argc, wchar_t** argv) {
  if (argc > 1 && wcscmp(argv[1], L"segment-overflow") == 0) {
    factor::cell page = factor::getpagesize();
    factor::segment segment(std::numeric_limits<factor::cell>::max() - page + 1,
                            false);
    return 2;
  }
  if (argc > 1 && wcscmp(argv[1], L"guard-overflow") == 0) {
    factor::segment segment(factor::getpagesize(), false,
                            std::numeric_limits<factor::cell>::max());
    return 2;
  }
  if (argc > 2 && wcscmp(argv[1], L"image-path") == 0) {
    factor::factor_vm* vm = new factor::factor_vm(NULL);
    const wchar_t* path = vm->default_image_path();
    check(wcscmp(path, argv[2]) == 0, "Default image path changed its directory");
    free((void*)path);
    return 0;
  }
  if (argc > 1 && wcscmp(argv[1], L"fatal-before-init") == 0) {
    // Another library may own TLS slot zero before Factor initializes TLS.
    DWORD key = TlsAlloc();
    check(key != TLS_OUT_OF_INDEXES, "TlsAlloc failed");
    check(TlsSetValue(key, (void*)1) != 0, "TlsSetValue failed");
    factor::fatal_error("Before VM initialization", 0);
    return 2;
  }
  if (argc > 1 && wcscmp(argv[1], L"fatal-after-init") == 0) {
    factor::init_mvm();
    factor::fatal_error("Before VM registration", 0);
    return 2;
  }
  // The compressed-image loader passes a narrow DLL name through this API.
  void* library = factor::native_dlopen("kernel32.dll");
  check(library != NULL, "native_dlopen failed to load a narrow DLL name");
  check(factor::native_dlsym(library, "GetCurrentProcessId") != NULL,
        "native_dlsym failed to resolve an export");
  factor::native_dlclose(library);
  factor::segment segment(factor::getpagesize(), false);
  MEMORY_BASIC_INFORMATION info;
  check(VirtualQuery((void*)segment.start, &info, sizeof(info)) == sizeof(info) &&
        info.Protect == PAGE_READWRITE, "Segment data is not writable");
  check(VirtualQuery((void*)(segment.start - 1), &info, sizeof(info)) == sizeof(info) &&
        info.Protect == PAGE_NOACCESS, "Segment low guard is not protected");
  check(VirtualQuery((void*)segment.end, &info, sizeof(info)) == sizeof(info) &&
        info.Protect == PAGE_NOACCESS, "Segment high guard is not protected");
  std::cout << "Windows service tests passed" << std::endl;
}

File: vm/tests/windows_services.py
#!/usr/bin/env python3
"""Windows VM startup diagnostics before a VM is available."""
import os
from pathlib import Path
import subprocess
import shutil
import tempfile
import unittest


@unittest.skipUnless(os.name == "nt", "Windows VM services")
class WindowsServicesTests(unittest.TestCase):
    def test_image_path_in_dotted_directory(self):
        executable = Path(__file__).resolve().parents[2] / "windows-services-tests.exe"
        with tempfile.TemporaryDirectory(suffix=".directory") as directory:
            for name in ("service.exe", "service"):
                with self.subTest(name=name):
                    copy = Path(directory) / name
                    shutil.copy2(executable, copy)
                    expected = str(Path(directory) / "service.image")
                    result = subprocess.run([str(copy), "image-path", expected],
                                            capture_output=True, text=True, timeout=10)
                    self.assertEqual(result.returncode, 0, result.stdout + result.stderr)

    def test_early_fatal_error(self):
        executable = Path(__file__).resolve().parents[2] / "windows-services-tests.exe"
        for mode, message in (("fatal-before-init", "Before VM initialization"),
                              ("fatal-after-init", "Before VM registration"),
                              ("segment-overflow", "Windows segment size overflow"),
                              ("guard-overflow", "Invalid Windows segment size")):
            with self.subTest(mode=mode):
                result = subprocess.run([str(executable), mode], capture_output=True,
                                        text=True, errors="replace", timeout=10)
                self.assertEqual(result.returncode, 1, result.stdout + result.stderr)
                self.assertIn("fatal_error: " + message, result.stderr)


if __name__ == "__main__":
    unittest.main()
